src/exchange/checkpoint.py

```diff
@@ -1,11 +1,63 @@
+from copy import deepcopy
+from typing import List
 from attrs import define, field
 
 
 @define
 class Checkpoint:
     """Checkpoint that counts the tokens in messages between the start and end index"""
 
-    start_index: int = field(default=-1)
-    end_index: int = field(default=0)
+    start_index: int = field(default=0)  # inclusive
+    end_index: int = field(default=0)  # inclusive
     token_count: int = field(default=0)
-    latest_generated_tokens: int = field(default=0)
+
+    def __deepcopy__(self, _) -> "Checkpoint":
+        """
+        Returns a deep copy of the Checkpoint object.
+        """
+        return Checkpoint(start_index=self.start_index, end_index=self.end_index, token_count=self.token_count)
+
+
+@define
+class CheckpointData:
```

Comment:
completely style, but as this is an aggregate, words that imply plural can help vs something like Data. At first, esp as there is a function about "message" in here, I thought there may be data, like messages in here..

```suggestion
class Checkpoints:
```

Reply:
Good idea!